### ISSUE TITLE
CpuAcc better (faster and better accuracy) than GpuAcc

### ISSUE THREAD

**Issue by ggardet** [ARM-software/armnn#304]
I am running ArmNN on HiKey960 with ARM Mali drivers and with `./mnist_tf_convol` (from https://github.com/ARM-software/Tool-Solutions/tree/master/ml-tool-examples/mnist-demo) I get better results (faster and better accuracy) on CpuAcc than on GpuAcc:
* CPU: `time ./mnist_tf_convol 1 1000` returns:
```
Optimisation mode: CpuAcc

[snip]
Prediction accuracy: 98.4%

real    0m5.020s
user    0m5.080s
sys     0m0.196s
```
* GPU: `time ./mnist_tf_convol 2 1000` returns:
```
Optimisation mode: GpuAcc

[snip]

Prediction accuracy: 90.9%

real    0m5.778s
user    0m1.672s
sys     0m0.921s
```

Is it somehow expected?

**Comment by MatthewARM** [comment]
Hmm no, that doesn't seem expected. The performance of the GPU on Hikey960 can be highly variable, but should average out faster than the CPU. And the results should be the same as CPU so the accuracy should be identical.

**Comment by ggardet** [comment]
Ok, so there is something wrong here. Would you have an HiKey 960 to try to reproduce this?

**Comment by MikeJKelly** [comment]
I have a HiKey 960 so I'll try to reproduce it here.

**Comment by ggardet** [comment]
CPU may be faster because it uses parallelization with 4 CPUs. Not sure if GPU has such a parallelization?
But anyway, IMO, the accuracy should be the same.

**Comment by MatthewARM** [comment]
The OpenCL execution is implicitly parallel - that's handled by the Compute Library and the OpenCL driver

**Comment by ggardet** [comment]
> The OpenCL execution is implicitly parallel - that's handled by the Compute Library and the OpenCL driver

So, there is definitely a problem somewhere.

**Comment by MatthewARM** [comment]
It would be helpful if you could post Arm NN profile output for a run on GPU and on CPU. I mean the json dump that you get with "event based profiling".

**Comment by ggardet** [comment]
> It would be helpful if you could post Arm NN profile output for a run on GPU and on CPU. I mean the json dump that you get with "event based profiling".

I have no idea how to do this. Would you have a guide for this, please?

**Comment by GeorgeARM** [comment]
Hello,

Is tuning enabled by the GpuAcc? Effective selection of LWS could be important for GPU execution.
Moreover, if the network is very small there is a slight possibility of CPU being faster as the GPU might not be able to fill up the cores.
Finally, concerning the accuracy, this might depend on what underlying algorithm is selected for performing the convolution layer but theoritically CpuAcc and GpuAcc should match.
Hope this helps.

**Comment by MatthewARM** [comment]
@ggardet I'm away from my desk right now but you do it with something like runtime->GetProfiler()->SetProfilingEnabled(true)

@GeorgeARM even if Winograd is selected on GPU for some reason, and not on CPU, I wouldn't expect to see an 8% accuracy drop on a reasonable model trained on the mnist data set. And on Hikey 960 even without tuning, in my experience the GPU should be about twice as fast as the CPU.

In short I'd really like to see profiling dumps from both backends - then we'd see exactly which kernels are being selected to ensure both backends are doing something sensible.

**Comment by james-conroy-arm** [comment]
Hi @ggardet ,

Did you manage to print out some profiling data as suggested by Matthew above?

Cheers,
James
